diff -Nur fontconfig-2.11.1-orig/src/fcxml.c fontconfig-2.11.1/src/fcxml.c
--- fontconfig-2.11.1-orig/src/fcxml.c	2014-03-13 14:00:47.000000000 +0700
+++ fontconfig-2.11.1/src/fcxml.c	2014-12-22 13:29:09.481284270 +0600
@@ -1322,6 +1322,7 @@
      * Have to swap the decimal point to match the current locale
      * if that locale doesn't use 0x2e
      */
+#ifndef __BIONIC__
     if ((dot = strchr (s, 0x2e)) &&
 	(locale_data = localeconv ()) &&
 	(locale_data->decimal_point[0] != 0x2e ||
@@ -1358,6 +1359,7 @@
 	}
     }
     else
+#endif
 	v = strtod (s, end);
     return v;
 }
