--- libtool-orig	2016-08-10 12:53:25.627658488 +0400
+++ libtool	2016-08-10 12:53:32.691880885 +0400
@@ -2525,8 +2525,8 @@
 	  # At present, this check doesn't affect windows .dll's that
 	  # are installed into $libdir/../bin (currently, that works fine)
 	  # but it's something to keep an eye on.
-	  test "$inst_prefix_dir" = "$destdir" && \
-	    func_fatal_error "error: cannot install \`$file' to a directory not ending in $libdir"
+	  #test "$inst_prefix_dir" = "$destdir" && \
+	  #    func_fatal_error "error: cannot install \`$file' to a directory not ending in $libdir"
 
 	  if test -n "$inst_prefix_dir"; then
 	    # Stick the inst_prefix_dir data into the link command.
