diff -Nur libXft-2.3.1-orig/src/xftglyphs.c libXft-2.3.1/src/xftglyphs.c
--- libXft-2.3.1-orig/src/xftglyphs.c	2012-06-02 23:36:35.000000000 +0700
+++ libXft-2.3.1/src/xftglyphs.c	2014-12-23 03:21:26.556212215 +0600
@@ -21,10 +21,10 @@
  */
 
 #include "xftint.h"
-#include <freetype/ftoutln.h>
-#include <freetype/ftlcdfil.h>
+#include FT_OUTLINE_H
+#include FT_LCD_FILTER_H
 
-#include <freetype/ftsynth.h>
+#include FT_SYNTHESIS_H
 
 /*
  * Validate the memory info for a font
