--- llvm-3.4/clang/lib/Driver/Tools.cpp.llvm.original	2014-07-01 10:11:13.998548162 +0400
+++ llvm-3.4/clang/lib/Driver/Tools.cpp	2014-07-01 10:35:41.526506430 +0400
@@ -6563,7 +6570,8 @@
   const SanitizerArgs &Sanitize = ToolChain.getSanitizerArgs();
   const bool IsPIE =
     !Args.hasArg(options::OPT_shared) &&
-    (Args.hasArg(options::OPT_pie) || Sanitize.hasZeroBaseShadow());
+    (Args.hasArg(options::OPT_pie) || Sanitize.hasZeroBaseShadow() ||
+     (isAndroid && !Args.hasArg(options::OPT_nopie)));
 
   ArgStringList CmdArgs;
 
